jsonutil/puritan/puritan.go

```diff
@@ -0,0 +1,340 @@
+		}
+	}
+	un.s = m
+
+	return nil
+}
+
+// Clean returns a Cleaned map[string]float64.
+func (un OldPriceMap) Clean() map[string]float64 {
+	return un.s
+}
+
+// PriceMap is a map used for prices.
+//
+// the keys are ISO currency codes; the values are strings of floats.
+type PriceMap struct {
+	s map[string]string
+}
+
+func (un *PriceMap) UnmarshalJSON(in []byte) error {
+	var m map[string]string
+	if err := json.Unmarshal(in, &m); err != nil {
+		return err
+	}
+	for k, v := range m {
+		if !okCurrency(k) {
+			return fmt.Errorf("invalid currency name %q", k)
+		}
+		if !okPrice(v) {
+			return fmt.Errorf("invalid price %q", v)
+		}
+	}
+	un.s = m
+
+	return nil
+}
+
+// Clean returns a map[string]string of cleaned Strings.
+func (un PriceMap) Clean() map[string]string {
+	return un.s
+}
+
+func okCurrency(s string) bool {
+	if len(s) != 3 {
+		return false
+	}
+	for _, r := range s {
+		if r < 'A' || r > 'Z' {
+			return false
+		}
+	}
+	return true
+}
+
+func okPrice(s string) bool {
+	// limits to what strconv.ParseFloat accepts
+	for _, r := range s {
+		switch r {
+		case '.', '+', '-', 'e', 'E':
+			// ok
+		default:
+			if r < '0' || r > '9' {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func u4(in []byte) rune {
+	if len(in) < 6 || in[0] != '\\' || in[1] != 'u' {
+		return -1
+	}
+	u, err := strconv.ParseUint(string(in[2:6]), 16, 32)
+	if err != nil {
+		return -1
+	}
+	return rune(u)
+}
+
+type uOpt struct {
+	nlok  bool
+	basic bool
+}
+
+func unmarshal(in []byte, o uOpt) (string, error) {
+	// heavily based on (inspired by?) unquoteBytes from encoding/json
+
+	if len(in) < 2 || in[0] != '"' || in[len(in)-1] != '"' {
+		// maybe it's a null and that's alright
+		if len(in) == 4 && in[0] == 'n' && in[1] == 'u' && in[2] == 'l' && in[3] == 'l' {
+			return "", nil
+		}
+		return "", fmt.Errorf("missing string delimiters: %q", in)
+	}
+
+	// prune the quotes
+	in = in[1 : len(in)-1]
+	i := 0
+	// try the fast track
+	for i < len(in) {
+		// 0x00..0x19 is the first of Cc
+		// 0x20..0x7e is all of printable ASCII (minus control chars)
+		if in[i] < 0x20 || in[i] > 0x7e || in[i] == '\\' || in[i] == '"' {
+			break
+		}
+		i++
+	}
+	if i == len(in) {
+		// wee
+		return string(in), nil
+	}
+	if o.basic {
+		return "", fmt.Errorf("invalid simple JSON string %q: not simple at %d", in, i)
+	}
+	// in[i] is the first problematic one
+	out := make([]byte, i, len(in)+2*utf8.UTFMax)
+	copy(out, in)
+	var r, r2 rune
+	var sz int
+	var c byte
+	var ubuf [utf8.UTFMax]byte
+	for i < len(in) {
+		switch c = in[i]; {
+		case c == '"':
+			return "", fmt.Errorf("unexpected unescaped quote at %d in \"%s\"", i, in)
+		case c < 0x20:
+			return "", fmt.Errorf("unexpected control character at %d in %q", i, in)
+		case c == '\\':
+			// handle escapes
+			i++
+			if i == len(in) {
+				return "", fmt.Errorf("unexpected end of string (trailing backslash) in \"%s\"", in)
+			}
+			switch in[i] {
+			case 'u':
```

Comment:
it this case a copy? or does it need careful reviewing?  it's glazing over prone at first pass